hamilflow/models/harmonic_oscillator.py

```diff
@@ -63,8 +64,112 @@ class HarmonicOscillatorIC(BaseModel):
+    for a [harmonic oscillator](https://en.wikipedia.org/wiki/Harmonic_oscillator).
+
+    :param system: all the params that defines the harmonic oscillator.
+    :param initial_condition: the initial condition of the harmonic oscillator.
+    """
+
+    def __init__(
+        self,
+        system: Dict[str, float],
+        initial_condition: Optional[Dict[str, float]] = {},
+    ):
+        self.system = HarmonicOscillatorSystem.model_validate(system)
+        self.initial_condition = HarmonicOscillatorIC.model_validate(initial_condition)
+
+    @cached_property
+    def definition(self) -> Dict[str, float]:
+        """model params and initial conditions defined as a dictionary."""
+        return {
+            "system": self.system.model_dump(),
+            "initial_condition": self.initial_condition.model_dump(),
+        }
+
+    @abstractmethod
+    def _x(self, t: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
+        r"""Solution to simple harmonic oscillators."""
+        ...
+
+    def __call__(self, n_periods: int, n_samples_per_period: int) -> pd.DataFrame:
+        """Generate time series data for the harmonic oscillator.
+
+        Returns a list of floats representing the displacement at each time step.
+
+        :param n_periods: Number of periods to generate.
+        :param n_samples_per_period: Number of samples per period.
+        """
+        time_delta = self.system.period / n_samples_per_period
+        time_steps = np.arange(0, n_periods * n_samples_per_period) * time_delta
+
+        data = self._x(time_steps)
+
+        return pd.DataFrame({"t": time_steps, "x": data})
+
+
+class SimpleHarmonicOscillator(HarmonicOscillatorBase):
+    r"""Generate time series data for a
+    [simple harmonic oscillator](https://en.wikipedia.org/wiki/Harmonic_oscillator).
+
+
+    In a one dimensional world, a mass $m$, driven by a force $F=-kx$, is described as
+
+    $$
+    \begin{align}
+    F &= - k x \\
+    F &= m a
+    \end{align}
+    $$
+
+    The mass behaves like a simple harmonic oscillator.
+
+    In general, the solution to a simple harmonic oscillator is
+
+    $$
+    x(t) = A \cos(\omega t + \phi),
+    $$
+
+    where $\omega$ is the angular frequency, $\phi$ is the initial phase, and $A$ is the amplitude.
+
+
+    To use this generator,
+
+    ```python
+    params = {"omega": omega}
+
+    ho = SimpleHarmonicOscillator(params=params)
+
+    df = ho(n_periods=1, n_samples_per_period=10)
+    ```
+
+    `df` will be a pandas dataframe with two columns: `t` and `x`.
+    """
+
+    def __init__(
+        self,
+        system: Dict[str, float],
+        initial_condition: Optional[Dict[str, float]] = {},
+    ):
+        super().__init__(system, initial_condition)
+        if self.system.type != "simple":
+            raise ValueError(
+                f"System is not a Simple Harmonic Oscillator: {self.system}"
+            )
+
+    def _x(self, t: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
```

Comment:
`Union[float, np.ndarray]` -> https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike

Reply:
Update in 01dc720

Reply:
https://github.com/kausalflow/hamilflow/pull/48/commits/01dc720e554a580dcd534424f2987a34b224d003